src/dir_set.ml

```diff
@@ -0,0 +1,196 @@
+open! Stdune
+
+module T = struct
+
+  type t = {
```

Comment:
Why do we have two records BTW? I'm also wondering if it be worth representing `Empty` and `Universal` as dedicated constructors in `t`

Reply:
Two records were just easier for me to think about because `children` can be thought of as a function [string -> t] and I expected some operations to work just on `children`.

Special representation of `Empty` and `Universal` is indeed not necessary, but I expected that to be a worthwhile optimization, given that we're testing things for whether they are Empty or Universal pretty often. They also have special status of being disallowed in certain contexts (I disallow the "useless" entries in the `exceptions` map).
Do you think we should remove the special representation?

Reply:
Do you mean add them back? It seems better to me to have them yes. Most operations match on it, and even for the one that don't it looks like it would still make them more efficient.

Reply:
I pushed this change.